shell/platform/linux/fl_key_event_plugin.cc

```diff
@@ -20,33 +39,258 @@ static constexpr char kUnicodeScalarValuesKey[] = "unicodeScalarValues";
+  self->id = id;
+  self->user_data = user_data;
+  return self;
+}
+
+// Calculates a unique ID for a given GdkEventKey object to use for
+// identification of responses from the framework.
+static uint64_t get_event_id(GdkEventKey* event) {
+  // Combine the event timestamp, the type of event, and the hardware keycode
+  // (scan code) of the event to come up with a unique id for this event that
+  // can be derived solely from the event data itself, so that we can identify
+  // whether or not we have seen this event already.
+  return (event->time & 0xffffffff) |
+         (static_cast<uint64_t>(event->type) & 0xffff) << 32 |
+         (static_cast<uint64_t>(event->hardware_keycode) & 0xffff) << 48;
+}
+
+// Finds an event in the event queue that was sent to the framework by its ID.
+static GdkEventKey* find_pending_event(FlKeyEventPlugin* self, uint64_t id) {
+  if (self->pending_events->len == 0 ||
+      FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->id != id) {
+    return nullptr;
+  }
+
+  return FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->event;
+}
+
+// Removes an event from the pending event queue.
+static void remove_pending_event(FlKeyEventPlugin* self, uint64_t id) {
+  if (self->pending_events->len == 0 ||
+      FL_KEY_EVENT_PAIR(g_ptr_array_index(self->pending_events, 0))->id != id) {
+    g_warning(
+        "Tried to remove pending event with id %ld, but the event was out of "
+        "order, or is unknown.",
+        id);
+    return;
+  }
+  g_ptr_array_remove_index(self->pending_events, 0);
+}
+
+// Adds an GdkEventKey to the pending event queue, with a unique ID, and the
+// plugin that added it.
+static void add_pending_event(FlKeyEventPlugin* self,
+                              uint64_t id,
+                              GdkEventKey* event) {
+  if (self->pending_events->len > kMaxPendingEvents) {
+    g_warning(
+        "There are %d keyboard events that have not yet received a "
+        "response from the framework. Are responses being sent?",
+        self->pending_events->len);
+  }
+  g_ptr_array_add(self->pending_events, fl_key_event_pair_new(id, event));
+}
+
+// Handles a response from the framework to a key event sent to the framework
+// earlier.
+static void handle_response(GObject* object,
+                            GAsyncResult* result,
+                            gpointer user_data) {
+  g_autoptr(FlKeyEventResponseData) data =
+      FL_KEY_EVENT_RESPONSE_DATA(user_data);
+
+  // Will also return if the weak pointer has been destroyed.
+  if (data->plugin == nullptr) {
+    return;
+  }
+
+  FlKeyEventPlugin* self = data->plugin;
+
+  g_autoptr(GError) error = nullptr;
+  FlBasicMessageChannel* messageChannel = FL_BASIC_MESSAGE_CHANNEL(object);
+  FlValue* message =
+      fl_basic_message_channel_send_finish(messageChannel, result, &error);
+  if (error != nullptr) {
+    g_error("Unable to retrieve framework response: %s", error->message);
```

Comment:
Should this be a `g_warning`? From the [docs](https://developer.gnome.org/glib/stable/glib-Message-Logging.html#g-error):
"Error messages are always fatal, resulting in a call to G_BREAKPOINT() to terminate the application. This function will result in a core dump; don't use it for errors you expect. Using this function indicates a bug in your program, i.e. an assertion failure."

Reply:
Switched to g_warning.